### PR TITLE
WebAuthn support

### LINKED ISSUE #68: [Tracking] WebAuthn support
Modify MANY to support WebAuthn requests

See https://webauthn.guide/#webauthn-api

### PR DESCRIPTION
**Pull request by fmorency**
## Description

Add WebAuthn COSE support

See https://webauthn.guide/#webauthn-api

## Related Issue

Fixes #68 

## Testing

- Albert + Yubikey + Ecdsa (Firefox)
- Albert + Yubikey + Eddsa (Chrome)
- `cargo test` in `many-rs`
- `cargo test` in `many-framework`
- Reftests in `many-protocol`

## Breaking Change

- [x] Yes
- [ ] No

Previous MANY clients using ECDSA are incompatible with this change. The ECDSA curve type was added to the `CoseKey` parameters which affect the `Identity`.

## Checklist:

- [x] I have read and followed the CONTRIBUTING guidelines for this project.
- [x] I have added or updated tests and they pass.
- [x] I have added or updated documentation and it is accurate.
- [x] I have noted any breaking changes in this module or downstream modules.

### CODE REVIEW

**Comment by fmorency** [comment]
@tantommy can you let me know here if this PR works with the frontend? Thanks!

**Comment by tantommy** [comment]
> @tantommy can you let me know here if this PR works with the frontend? Thanks!

confirmed, this is working in desktop firefox on macOS as well

**Comment by fmorency** [comment]
@hansl @tantommy Please review this PR! Thanks!

**Comment by fmorency** [comment]
Re-converting this to Draft.

I don't think we're using WebAuthn properly at the moment. We need to discuss this.

File: src/many/src/cose_helpers.rs [hunk 1]
```diff
@@ -12,10 +12,16 @@ use std::collections::{BTreeMap, BTreeSet};
 /// * `x` - Public key
 /// * `d` - Private key
 pub fn eddsa_cose_key(x: Vec<u8>, d: Option<Vec<u8>>) -> CoseKey {
-    let mut params: Vec<(Label, Value)> = Vec::from([(
-        Label::Int(coset::iana::OkpKeyParameter::X as i64),
-        Value::Bytes(x),
-    )]);
+    let mut params: Vec<(Label, Value)> = Vec::from([
```
Comment (hansl) [inline src/many/src/cose_helpers.rs:15]: I think we should move that change in a separate PR ASAP so people will get their new ID quick.

File: src/many/src/message.rs [hunk 2]
```diff
@@ -181,24 +186,131 @@ impl CoseSign1RequestMessage {
...
+    /// See https://webauthn.guide/#webauthn-api
+    fn _verify_webauthn(
+        &self,
+        unprotected: BTreeMap<Label, Value>,
+        key: CoseKeyIdentity,
+    ) -> Result<(), String> {
+        #[derive(Deserialize)]
+        struct ClientData {
+            challenge: String,
+            #[allow(dead_code)]
```
Comment (hansl) [inline src/many/src/message.rs:201]: Please remove to make sure we actually use the origin

File: src/many/src/message.rs [hunk 3]
```diff
@@ -181,24 +186,131 @@ impl CoseSign1RequestMessage {
...
+    ///
+    /// This is non-standard COSE
+    /// See https://webauthn.guide/#webauthn-api
+    fn _verify_webauthn(
+        &self,
+        unprotected: BTreeMap<Label, Value>,
+        key: CoseKeyIdentity,
+    ) -> Result<(), String> {
+        #[derive(Deserialize)]
+        struct ClientData {
```
Comment (hansl) [inline src/many/src/message.rs:199]: Can't you get this type from a crate? Seems like the webauthn-rs crate could be used; https://docs.rs/webauthn-rs/latest/webauthn_rs/proto/struct.CollectedClientData.html

File: src/many/src/message.rs (hunk 2 above)
Comment (fmorency) [reply src/many/src/message.rs:201]: Good catch. We should add a clippy directive to catch those if it exists.

File: src/many/src/message.rs (hunk 3 above)
Comment (fmorency) [reply src/many/src/message.rs:199]: I can. However, is it really worth it to bring in this crate (only) for this?

File: src/many/src/cose_helpers.rs (hunk 1 above)
Comment (fmorency) [reply src/many/src/cose_helpers.rs:15]: Can do.

Comment (fmorency) [reply src/many/src/cose_helpers.rs:15]: See #86

**Comment by fmorency** [comment]
@hansl I had another PR opened (#84), but you reviewed this one; my bad. Your review is still relevant, however. I merged my other PR in this branch.

You can review it again.

File: src/many/src/server.rs [hunk 4]
```diff
@@ -21,6 +24,10 @@ pub struct ManyModuleList {}
 
 pub const MANYSERVER_DEFAULT_TIMEOUT: u64 = 300;
 
+/// List of allowed URLs to communicate with the server
+/// WebAuthn-only
+pub static ALLOWED_URLS: OnceCell<Option<Vec<ManyUrl>>> = OnceCell::new();
```
Comment (hansl) [inline src/many/src/server.rs:29]: Oh my god. I think we can do better.

Comment (hansl) [reply src/many/src/server.rs:29]: At least `thread_local`...

**Review by hansl — CHANGES_REQUESTED** [review]
Just need at least a couple of tests. Otherwise LGTM

**Comment by fmorency** [comment]
We have 100% percent code coverage for this feature.

**Review by hansl — APPROVED** [review]
